estransport/connection.go

```diff
@@ -0,0 +1,280 @@
+// Licensed to Elasticsearch B.V. under one or more agreements.
+// Elasticsearch B.V. licenses this file to you under the Apache 2.0 License.
+// See the LICENSE file in the project root for more information.
+
+package estransport
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"net/url"
+	"sort"
+	"sync"
+	"time"
+)
+
+var (
+	defaultResurrectTimeoutInitial      = 60 * time.Second
+	defaultResurrectTimeoutFactorCutoff = 5
+)
+
+// ConnectionPool defines the interface for the connection pool.
+//
+type ConnectionPool interface {
+	Next() (*Connection, error)
+	Remove(*Connection) error
```

Comment:
what is the purpose of Remove? Is a connection to be removed forever?

Is removed only after failure? If so, how about naming it failed?

Utilities like ConnectionPools are sensitive to the protocol. For example: do we need to full consumer if a connection failed? Do we have to close the underlying TCP connection if the connection failed? Not properly documenting the protocol and required behavior. Mixing wrong implementations can easily lead to security issues. Even better if there is a common connection pool protocol test suite.

A good read of potential issues, and how the custom connection pool in the redis golang library did lead to a security breah: https://medium.com/@cep21/interface-contracts-matter-812dd43709b3